src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -158,40 +158,134 @@ bool Lowering::IsContainableImmed(GenTree* parentNode, GenTree* childNode) const
+
+    // We cannot contain if the parent node
+    // * is contained
+    // * is not operating on an integer
+    // * is already marking a child node as contained
+    // * is required to throw on overflow
+
     if (parentNode->isContained())
         return false;
 
     if (!varTypeIsIntegral(parentNode))
         return false;
 
-    if (parentNode->gtFlags & GTF_SET_FLAGS)
+    if (parentNode->gtGetOp1()->isContained() || parentNode->gtGetOp2()->isContained())
         return false;
 
-    GenTree* op1 = parentNode->gtGetOp1();
-    GenTree* op2 = parentNode->gtGetOp2();
+    if (parentNode->OperMayOverflow() && parentNode->gtOverflow())
+        return false;
 
-    if (op2 != childNode)
+    // We cannot contain if the child node:
+    // * is not operating on an integer
+    // * is required to set a flag
+    // * is required to throw on overflow
+
+    if (!varTypeIsIntegral(childNode))
         return false;
 
-    if (op1->isContained() || op2->isContained())
+    if ((childNode->gtFlags & GTF_SET_FLAGS) != 0)
         return false;
 
-    if (!varTypeIsIntegral(op2))
+    if (childNode->OperMayOverflow() && childNode->gtOverflow())
         return false;
 
-    if (op2->gtFlags & GTF_SET_FLAGS)
+    GenTree* matchedOp = nullptr;
+
+    if (childNode->OperIs(GT_MUL))
+    {
+        if (childNode->gtGetOp1()->isContained() || childNode->gtGetOp2()->isContained())
+        {
+            // Cannot contain if either of the childs operands is already contained
+            return false;
+        }
+
+        if ((parentNode->gtFlags & GTF_SET_FLAGS) != 0)
+        {
+            // Cannot contain if the parent operation needs to set flags
+            return false;
+        }
+
+        if (parentNode->OperIs(GT_ADD))
+        {
+            // Find "c + (a * b)" or "(a * b) + c"
+            return IsSafeToContainMem(parentNode, childNode);
+        }
+
+        if (parentNode->OperIs(GT_SUB))
+        {
+            // Find "c - (a * b)"
+            assert(childNode == parentNode->gtGetOp2());
+            return IsSafeToContainMem(parentNode, childNode);
+        }
+
+        // TODO: Handle mneg
         return false;
+    }
 
-    // Find "a + b * c" or "a - b * c".
-    if (parentNode->OperIs(GT_ADD, GT_SUB) && op2->OperIs(GT_MUL))
+    if (childNode->OperIs(GT_LSH, GT_RSH, GT_RSZ))
     {
-        if (parentNode->gtOverflow())
+        // Find "a op (b shift cns)"
+
+        if (childNode->gtGetOp1()->isContained())
+        {
+            // Cannot contain if the childs op1 is already contained
+            return false;
+        }
+
+        GenTree* shiftAmountNode = childNode->gtGetOp2();
+
+        if (!shiftAmountNode->IsCnsIntOrI())
+        {
+            // Cannot contain if the childs op2 is not a constant
+            return false;
+        }
+
+        target_ssize_t shiftAmount = (target_ssize_t)shiftAmountNode->AsIntCon()->gtIconVal;
```

Comment:
Why the `target_ssize_t` cast (doesn't look necessary)?
```suggestion
        target_ssize_t shiftAmount = (target_ssize_t)shiftAmountNode->AsIntCon()->IconValue();
```

Reply:
I had just copied the line from `IsContainableImmed`, it'd be because they call emitter APIs which require `target_ssize_t` and the result of `gtIconVal` is `ssize_t`.

We're not doing that here since we have other requirements for the check (regular shift containment also doesn't go through `isContainableImmed`, although it maybe should).

```suggestion
        ssize_t shiftAmount = shiftAmountNode->AsIntCon()->IconValue();
```